dvc/command/pipeline.py

```diff
@@ -162,13 +162,18 @@ def run(self):
 
 class CmdPipelineList(CmdBase):
     def run(self):
-        import networkx
+        import networkx as nx
 
         pipelines = self.repo.pipelines
         for p in pipelines:
-            stages = networkx.get_node_attributes(p, "stage")
-            for stage in stages:
-                logger.info(stage)
+            stages = nx.get_node_attributes(p, "stage")
+            for node in p:
```

Comment:
Please correct me if I'm wrong, but what you are doing here is you are walking through all nodes in the pipeline in the random order, then you do a dfs_postorder for each one of those and then just print it. This is basically the same as what it was before but now with a redundant step of `dfs_postorder_nodes`. :slightly_smiling_face: Looks like we can do better than that :slightly_smiling_face:

Reply:
Good way to think about it is to imagine how the pipeline looks. So it has a N start nodes (`p.out_degree(node) == 0`), M end nodes (`p.in_degree(node) == 0`) and some nodes between them. The way `dvc repro --pipeline` works, is it finds all end nodes and for each of them does `dfs_postorder_nodes` and just runs those. But obviously, that real order of execution is not going to look good here :slightly_frowning_face:

Reply:
Thank you for the explanation. I'll work on it.

Reply:
`p.out_degree(node) == 0` for end nodes and `p.in_degree(node) == 0` for start nodes right?